File: app/src/main/java/com/robifr/ledger/ui/main/MainActivity.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.main;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.util.TypedValue;
import android.view.MenuItem;
import android.view.View;
import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.text.HtmlCompat;
import androidx.navigation.NavController;
import androidx.navigation.NavDestination;
import androidx.navigation.NavOptions;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.NavHostFragment;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.navigation.NavigationBarView;
import com.robifr.ledger.R;
import com.robifr.ledger.databinding.MainActivityBinding;
import com.robifr.ledger.util.Compats;
import dagger.hilt.android.AndroidEntryPoint;
import java.util.List;
import java.util.Objects;

@AndroidEntryPoint
public class MainActivity extends AppCompatActivity
    implements NavigationBarView.OnItemSelectedListener,
        NavController.OnDestinationChangedListener {
  @Nullable private MainActivityBinding _activityBinding;
  @Nullable private MainCreate _create;
  @Nullable private MainResultHandler _resultHandler;

  @Override
  public boolean onNavigationItemSelected(@NonNull MenuItem item) {
    Objects.requireNonNull(item);
    Objects.requireNonNull(this._activityBinding);

    final NavController navController = Navigation.findNavController(this, R.id.fragmentContainer);
    final NavDestination currentDestination =
        Objects.requireNonNull(navController.getCurrentDestination());

    navController.navigate(
        item.getItemId(),
        null,
        new NavOptions.Builder().setPopUpTo(currentDestination.getId(), true).build());
    return true;
  }

  @Override
  public void onDestinationChanged(
      @NonNull NavController navController,
      @NonNull NavDestination destination,
      @Nullable Bundle bundle) {
    Objects.requireNonNull(navController);
    Objects.requireNonNull(destination);
    Objects.requireNonNull(this._activityBinding);

    // Match selected item of bottom navigation with the visible fragment.
    // It doesn't get matched on back pressed.
    final MenuItem item =
        this._activityBinding.bottomNavigation.getMenu().findItem(destination.getId());
    if (item != null) item.setChecked(true);

    final int bottomNavigationVisibility =
        destination.getParent() != null && destination.getParent().getId() == R.id.mainGraph
            ? View.VISIBLE
            : View.GONE;
    final int createButtonVisibility =
        destination.getParent() != null
                && destination.getParent().getId() == R.id.mainGraph
                && destination.getId() != R.id.dashboardFragment
            ? View.VISIBLE
            : View.GONE;
    final TypedValue backgroundColor = new TypedValue();
    final int navigationBarColor =
        bottomNavigationVisibility == View.VISIBLE
            ? com.google.android.material.R.attr.colorSurface
            : android.R.attr.colorBackground;

    // Due to bottom navigation on main activity uses a different color — color surface.
    // Match system navigation bar color into it, otherwise with current background.
    this.getTheme().resolveAttribute(navigationBarColor, backgroundColor, true);
    this.getWindow().setNavigationBarColor(this.getColor(backgroundColor.resourceId));

    // Hide views on main activity when user navigating to another fragment other than
    // the one defined as top of the stack — queue, customer, and product — inside bottom
    // navigation.
    this._activityBinding.bottomNavigation.setVisibility(bottomNavigationVisibility);
    this._activityBinding.createButton.setVisibility(createButtonVisibility);
  }

  @Override
  protected void onCreate(@Nullable Bundle savedInstance) {
    super.onCreate(savedInstance);

    this._activityBinding = MainActivityBinding.inflate(this.getLayoutInflater());
    this.setContentView(this._activityBinding.getRoot());

    this._create = new MainCreate(this);
    this._resultHandler = new MainResultHandler(this);

    this._activityBinding.createButton.setOnClickListener(button -> this._create.openDialog());
    this._activityBinding.bottomNavigation.setOnItemSelectedListener(this);
    this.getOnBackPressedDispatcher().addCallback(this, new OnBackPressedHandler());
    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);

    List.of(R.id.dashboardFragment, R.id.queueFragment, R.id.customerFragment, R.id.productFragment)
        .forEach(
            id ->
                this._activityBinding
                    .bottomNavigation
                    .findViewById(id)
                    .setOnLongClickListener(Compats::hideTooltipText));

    final NavHostFragment navHostFragment =
        (NavHostFragment)
            Objects.requireNonNull(
                this.getSupportFragmentManager().findFragmentById(R.id.fragmentContainer));
    navHostFragment.getNavController().addOnDestinationChangedListener(this);

    if (!Environment.isExternalStorageManager()) this.requireStoragePermission();
  }

  @NonNull
  public MainActivityBinding activityBinding() {
    return Objects.requireNonNull(this._activityBinding);
  }

  @NonNull
  public Intent requireStoragePermission() {
    Objects.requireNonNull(this._resultHandler);

    final Intent intent =
        new Intent(
            Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION,
            Uri.fromParts("package", this.getPackageName(), null));

    new MaterialAlertDialogBuilder(this)
        .setTitle(this.getString(R.string.text_storage_access_permission_required))
        .setMessage(
            HtmlCompat.fromHtml(
                this.getString(R.string.maindialog_manageexternalstorage_permission_description),
                HtmlCompat.FROM_HTML_MODE_LEGACY))
        .setNegativeButton(
            this.getString(R.string.text_deny_and_quit), (dialog, type) -> this.finish())
        .setPositiveButton(
            this.getString(R.string.text_grant),
            (dialog, type) -> this._resultHandler.permissionLauncher().launch(intent))
        .setCancelable(false)
        .show();
    return intent;
  }

  private class OnBackPressedHandler extends OnBackPressedCallback {
    public OnBackPressedHandler() {
      super(true);
    }

    @Override
    public void handleOnBackPressed() {
      MainActivity.this.finish();
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/main/MainActivity.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.main

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.graphics.Insets
import androidx.core.os.LocaleListCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.NavOptions
import androidx.navigation.findNavController
import com.google.android.material.navigation.NavigationBarView
import com.robifr.ledger.R
import com.robifr.ledger.databinding.MainActivityBinding
import com.robifr.ledger.ui.settings.AppUpdateAvailable
import com.robifr.ledger.ui.settings.viewmodel.SettingsDialogState
import com.robifr.ledger.ui.settings.viewmodel.SettingsViewModel
import com.robifr.ledger.ui.settings.viewmodel.UnknownSourceInstallationDialogState
import com.robifr.ledger.ui.settings.viewmodel.UpdateAvailableDialogState
import com.robifr.ledger.util.hideTooltipText
import dagger.hilt.android.AndroidEntryPoint
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

@AndroidEntryPoint
open class MainActivity :
    AppCompatActivity(),
    NavigationBarView.OnItemSelectedListener,
    NavController.OnDestinationChangedListener,
    ActivityResultCallback<ActivityResult> {
  private var _activityBinding: MainActivityBinding? = null
  val activityBinding: MainActivityBinding
    get() = _activityBinding!!

  private val _settingsViewModel: SettingsViewModel by viewModels()
  protected lateinit var _permission: RequiredPermission
  private lateinit var _permissionLauncher: ActivityResultLauncher<Intent>
  private lateinit var _create: MainCreate

  override fun onNavigationItemSelected(item: MenuItem): Boolean {
    findNavController(R.id.fragmentContainer).apply {
      currentDestination?.id?.let {
        navigate(item.itemId, null, NavOptions.Builder().setPopUpTo(it, true).build())
      }
    }
    return true
  }

  override fun onDestinationChanged(
      controller: NavController,
      destination: NavDestination,
      arguments: Bundle?
  ) {
    // Match selected item of bottom navigation with the visible fragment.
    // It doesn't get matched on back pressed.
    activityBinding.bottomNavigation.menu.findItem(destination.id)?.isChecked = true
    // Hide views on main activity when user navigating to another fragment other than the one
    // defined as top of the stack — queue, customer, and product — inside bottom navigation.
    activityBinding.bottomNavigation.isVisible = destination.parent?.id == R.id.mainGraph
    activityBinding.createButton.isVisible =
        destination.parent?.id == R.id.mainGraph && destination.id != R.id.dashboardFragment
  }

  override fun onActivityResult(result: ActivityResult) {
    finish()
    startActivity(intent)
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    enableEdgeToEdge()
    super.onCreate(savedInstanceState)
    _activityBinding = MainActivityBinding.inflate(layoutInflater)
    setContentView(activityBinding.root)

    _permission = RequiredPermission(this)
    _create = MainCreate(this)
    activityBinding.bottomNavigation.setOnItemSelectedListener(this)
    ViewCompat.setOnApplyWindowInsetsListener(activityBinding.bottomNavigation) { view, insets ->
      val windowInsets: Insets = insets.getInsets(WindowInsetsCompat.Type.navigationBars())
      view.updatePadding(bottom = windowInsets.bottom)
      WindowInsetsCompat.CONSUMED
    }
    AppCompatDelegate.setApplicationLocales(
        LocaleListCompat.forLanguageTags(
            _settingsViewModel.uiState.safeValue.languageUsed.languageTag))
    listOf(R.id.dashboardFragment, R.id.queueFragment, R.id.customerFragment, R.id.productFragment)
        .forEach { activityBinding.bottomNavigation.findViewById<View>(it)?.hideTooltipText() }

    _permissionLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult(), this)
    if (!_permission.isStorageAccessGranted()) {
      _permission.showStorageAccessDialog(
          onDeny = { finish() },
          onGrant = { _permissionLauncher.launch(_permission.storageAccessIntent()) })
    } else {
      // Only automatically check for app update once a day.
      if (_settingsViewModel.uiState.safeValue.isLastCheckedTimeForAppUpdatePastMidNight()) {
        _settingsViewModel.dialogState.observe(this, ::_onSettingsDialogState)
        _settingsViewModel.onCheckForAppUpdate(this, false)
      }
    }
  }

  override fun onStart() {
    super.onStart()
    findNavController(R.id.fragmentContainer).addOnDestinationChangedListener(this)
  }

  private fun _onSettingsDialogState(state: SettingsDialogState) {
    when (state) {
      is UpdateAvailableDialogState -> {
        val dateFormat: DateTimeFormatter =
            DateTimeFormatter.ofPattern(
                getString(_settingsViewModel.uiState.safeValue.languageUsed.fullDateFormat))
        AppUpdateAvailable(this)
            .showDialog(
                updateVersion = state.githubRelease.tagName,
                updateDate =
                    ZonedDateTime.parse(
                            state.githubRelease.publishedAt, DateTimeFormatter.ISO_DATE_TIME)
                        .format(dateFormat),
                updateSize = state.githubRelease.sizeInMb(),
                onUpdate = { _settingsViewModel.onUpdateApp() })
      }
      is UnknownSourceInstallationDialogState -> {
        _permission.showUnknownSourceInstallationDialog {
          _permissionLauncher.launch(_permission.unknownSourceInstallationIntent())
        }
      }
    }
  }
}
